Qsvenc enhancement

lizhong1008 opened the pull request.

uartie commented:
Have these been submitted to upstream ML, yet?  If so, please provide link(s).

dvrogozh commented on libavcodec/qsvenc.c:
```diff
@@ -494,10 +500,10 @@ static int init_video_param(AVCodecContext *avctx, QSVEncContext *q)
-    q->width_align = avctx->codec_id == AV_CODEC_ID_HEVC ? 32 : 16;
+    // If the minor version is greater than or equal to 19,
+    // then can use the same alignment settings as H.264 for HEVC
+    q->width_align = (avctx->codec_id != AV_CODEC_ID_HEVC ||
```
Where this requirement was taken from? Mediasdk requirement is: "Width must be a multiple of 16. Height must be a multiple of 16 for progressive frame sequence and a multiple of 32 otherwise."

To me what is programmed here looks completely incorrect.

lizhong1008 replied:
I guess some old version of MSDK weren't totally  following the documentation you mentioned. 
width_alignment of hevc is 32 from the beginning of Luca's enabling: https://code.oxygene.sk/mirrors/ffmpeg/commit/66acb76bb0492b263215ca9b4d927a7be39ace02#b8b2636776a36ebc6cd1e21f1ec8d7dfac28dc6f_52_52 
And Aaron added such a comment in  ae5b67ee64bec00c52881ff8426fddaf477341c1 
To confirm this, we may need to check some old version MSDK. 
But anyway, for current MSDK, even for hevc encoding, 16 bytes is enough for width and this is the patch is doing. (I admit that I am conservative that don't change to 16 bytes totally to keep compatibility, because it is not easy to check old version MSDK for me and I chose to believe Luca and Aaron them definitely found 16 bytes aligned was not enough for old version MSDK).
BTW,  comment is always welcome, but should be better discuss in ffmpeg-devel.

lizhong1008 replied:
And also, if it was me, if I found hevc encoding was not following the documentation, I would never give a workaround patch in ffmpeg. Instead, I would like to ask MSDK team to fix it. 
However, it has already happened, I prefer to keep compatibility with both old version ffmpeg and old version MSDK.  I don't want to take the risk to see any complain (who is still using some old MSDK) that why old version ffmpeg can work well but broken right now? I don't want to explain the technological detail and then force him to update MSDK/MSS (you know, updating MSS version needs to update centos version and Linux kernel version as well).

lizhong1008 commented:
@uartie, yes, you can find all my patches on: https://patchwork.ffmpeg.org/project/ffmpeg/list/?submitter=387&state=*&q=&archive=&delegate=  
Patchwork just missed two of them, I don't know why. And cover-letter hasn't be received  there.

uartie commented:
FATE tests have been enabled in `intel/ci/ubuntu/bionic/pr/check`.  Please rebase this pull-request for latest FATE patch to fix compile issue.

uartie commented:
> FATE tests have been enabled in `intel/ci/ubuntu/bionic/pr/check`. Please rebase this pull-request for latest FATE patch to fix compile issue.

Fixed.  It was a CI configuration issue.

lizhong1008 committed:
lavc/qsvenc: add BRC sliding window setting

WinBRCMaxAvgKbps is to specify maximum average bitrate over a
sliding window with size of WinBRCSize

WinBRCMaxAvgKbps will be ignored in CBR mode and equal to TargetKbps.

Signed-off-by: Zhong Li <zhong.li@intel.com>

uartie commented:
@lizhong1008 please rebase to capture merged patches and test fixes.

lizhong1008 commented:
@uartie , rebased.

lizhong1008 committed:
lavc/qsvenc: enable hevc gpb option

GPB is the default type, just contains forward references but the
slice_type is B slice.
Add this option to allow user to set regular P slice.

Fix ticket#6870

Signed-off-by: Zhong Li <zhong.li@intel.com>

uartie commented:
@lizhong1008 have all these been merged upstream, yet.  If so, please close.  Otherwise, maybe rebase

lizhong1008 commented:
@uartie ,  not all have been merged. And I think we need to save time of do rebase work.  We should rebase when and only when new version patch has been update.

uartie commented:
3 out of 5 of these patches have been merged upstream.  Also, our tests have been updated to specify rc_mode option in ffmpeg commands.  The rc_mode option requires upstream commit f4d7bd6cdbcf.  So please rebase this patch series to resolve conflicts and applied on top of latest master branch.

lizhong1008 commented:
@uartie, it has been rebased.

lizhong1008 committed:
configure: include pkgconfig path as vaapi header search

Currectly just standard header path can be found,
check_type/struct will fail if vaapi is installed somewhere else.

Move them followed "check_pkg_config"

Reviewed-by: Mark Thompson <sw@jkqxz.net>
Reviewed-by: Timo Rothenpieler <timo@rothenpieler.org>
Signed-off-by: Zhong Li <zhong.li@intel.com>

lizhong1008 committed:
lavc/hevc_ps_enc: fix vps nal issues

1. write vps_num_hrd_parameters element
2. flush left bits

Signed-off-by: Zhong Li <zhong.li@intel.com>

lizhong1008 committed:
lavc/qsvenc: fix hevc vps extradata issues

cbs trace qsv vps header failed due to some reasons:
1. vps_temporal_id_nesting_flag is not set but spec required it must to
   be 1 when vps_max_sub_layers_minus1 is equal to 0.
2. vps_num_hrd_parameters is not set and written.
3. other issues in ff_hevc_encode_nal_vps() (have fixed in pervious commit).

Reproduce: ffmpeg -hwaccel qsv -v verbose -c:v h264_qsv -i bbb_sunflower_1080p_30fps_normal.mp4 -vframes 1
-c:v hevc_qsv  -bsf:v trace_headers -f null -

Signed-off-by: Zhong Li <zhong.li@intel.com>

lizhong1008 committed:
lavc/qsvenc: make the queried libmfx version easily reused

Signed-off-by: Zhong Li <zhong.li@intel.com>

lizhong1008 committed:
lavc/qsvenc: get vps extradata from MSDK

Signed-off-by: Zhong Li <zhong.li@intel.com>

lizhong1008 committed:
lavc/qsvenc: enable hevc coding options configuration

Signed-off-by: Zhong Li <zhong.li@intel.com>

lizhong1008 committed:
qsv: fix the dangerous macro definitions

Signed-off-by: Zhong Li <zhong.li@intel.com>

lizhong1008 committed:
lavc/qsvenc: use the common option "trellis" of AVCodecContext

Signed-off-by: Zhong Li <zhong.li@intel.com>

lizhong1008 committed:
lavc/qsvenc_h264: remove the privite option trellis

There are many problems of current qsv trellis option:
1. Duplicated with AVCodecContext definition
2. MFX_TRELLIS_XXX is introduced by MSDK API 1.17
   Currently Without MSDK API checking thus may cause compilation issue.
3. user is inclined to enable trellis when set "-trellis 1", but
   actually it is to disable since MFX_TRELLIS_OFF is equal to 1.
4. It is too complex for user to enable trellis for every frame(I/P/B).

Just simply remove the private option, and switch to the AVCodecContext
definition. Compatibility should not a big problem (expect can't exact map)
since the option name is same as AVCodecContext.

Signed-off-by: Zhong Li <zhong.li@intel.com>

uartie commented:
Closing.  Many of these patches have been merged.  Feel free to open a new PR with narrower scope for remaining patches.
